### ISSUE TITLE
Allow AI SDK usage without requiring Zod Installation

### ISSUE THREAD

**Issue by anteriovieira** [vercel/ai#3282]
### Description

Currently, the `vercel/ai` SDK requires Zod as a peer dependency for internal validation tasks. However, many projects, like ours, use different validation libraries such as Joi and do not utilize Zod directly in the main code. Including Zod in our package.json solely to use the vercel/ai package is confusing for users who do not see it being used in the codebase.

To enhance flexibility and reduce unnecessary dependencies, we propose making the `vercel/ai` SDK more accommodating of different validation libraries or providing a way to use the SDK without installing Zod. This change would prevent versioning issues and make the SDK more user-friendly for projects that do not rely on Zod.

Would it be possible to investigate and implement a solution that allows the use of the `vercel/ai `SDK without requiring Zod? 

### Additional context

I'm using the AI package in a project where we use [Joi](https://joi.dev/api/?v=17.13.3) for validation instead of Zod. So I've created a custom joiSchema function to use with the `generateObject` feature:

```ts
// ai/utils.ts
import { jsonSchema } from 'ai'
import type { Schema } from 'ai'
import * as parse from 'joi-to-json'
import * as j from 'joi'

const joiToJsonSchema = parse as any

export function joiSchema<OBJECT>(schema: j.Schema<OBJECT>): Schema<OBJECT> {
    return jsonSchema(
        joiToJsonSchema(schema),
        {
            validate: value => {
                const result = schema.validate(value)

                return result.error
                    ? {
                        success: false,
                        error: result.error,
                    }
                    : {
                        success: true,
                        value: result.value as OBJECT,
                    }
            },
        }
    )
};
```

```ts
// using joiSchema
const { object } = await generateObject<T>({
    schema: joiSchema(myJoiSchema),
    ...
})
```

I noticed that the package has Zod listed as an optional peerDependency in the package.json:

```json
"peerDependenciesMeta": {
  "zod": {
    "optional": true
  },
  ...
},
```

However, when trying to use the `generateObject` feature without Zod installed, I encountered the following error:

```
Error: Cannot find module 'zod'
Require stack:
- /Users/.../node_modules/zod-to-json-schema/dist/cjs/parseDef.js
- /Users/.../node_modules/zod-to-json-schema/dist/cjs/index.js
- /Users/.../node_modules/@ai-sdk/ui-utils/dist/index.js
- /Users/.../node_modules/ai/dist/index.js
// ... (full stack trace)
```

It appears that the zod-to-json-schema module, a dependency of `@ai-sdk/ui-utils`, is trying to import Zod even when it's not installed.

_Originally posted https://github.com/vercel/ai/discussions/3160_

**Comment by cloudkite** [comment]
Would love this as well, mainly as a way to keep bundle size down as we are deploying to edge functions

**Comment by austinm911** [comment]
Standard Schema is now 1.0, so perhaps the library could move to this

https://standardschema.dev/

**Comment by lgrammel** [comment]
> Standard Schema is now 1.0, so perhaps the library could move to this, although I know there are complexities associated with converted to JSON schema that is currently only handled with zod.
> 
> https://standardschema.dev/

No, because standard schema does not cover schema to JSON schema conversion.

**Comment by eric-burel** [comment]
For googlers: it means that if you hit `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'zod'` you probably just need to install `zod`.

**Comment by aayush-kapoor** [comment]
Marking this as closed for now - if you face any regression or have a similar problem after migrating to the latest AI SDK version, please feel free to open a new issue (with possible reproduction)

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This describes a concrete dependency-loading failure: an optional peer dependency is effectively required even when the caller supplies a non-Zod schema. Although broader validation-library flexibility is requested, the existing unexpected runtime failure makes this a bug.

### Evidence
- Zod is declared as an optional peer dependency, but using `generateObject` with a custom Joi-backed `Schema` fails when Zod is absent.
- The reported stack trace shows `zod-to-json-schema` importing the uninstalled `zod` package through `@ai-sdk/ui-utils`.
- The reporter provided code that avoids Zod yet still encounters `Cannot find module 'zod'`; the closing comment suggests opening a new issue if this behavior regresses on the latest SDK.
